src/realm/sync/network/default_socket.cpp

```diff
@@ -362,9 +365,153 @@ void DefaultWebSocketImpl::initiate_websocket_handshake()
 }
 } // namespace
 
+DefaultSocketProvider::DefaultSocketProvider(const std::shared_ptr<util::Logger>& logger,
+                                             const std::string user_agent)
+    : m_logger_ptr{logger}
+    , m_service{}
+    , m_random{}
+    , m_user_agent{user_agent}
+    , m_state_mutex{}
+    , m_state{NotStarted}
+    , m_thread{}
+    , m_stop_future{}
+{
+    REALM_ASSERT(m_logger_ptr);                     // Make sure the logger is valid
+    util::seed_prng_nondeterministically(m_random); // Throws
+    start_keep_running_timer();                     // TODO: Update service so this timer is not needed
+}
+
+DefaultSocketProvider::~DefaultSocketProvider()
+{
+    // Wait for the thread to stop
+    stop(true);
+    if (m_keep_running_timer)
+        m_keep_running_timer->cancel();
+    if (m_thread.joinable()) {
+        m_thread.join();
+    }
+    REALM_ASSERT(m_state == Stopped || m_state == NotStarted);
+}
+
+bool DefaultSocketProvider::start()
+{
+    std::unique_lock<std::mutex> lock{m_state_mutex};
+    // Start the thread if it hasn't been started or is stopped
+    if (m_state != NotStarted && m_state != Stopped) {
+        return false;
+    }
+
+    // If the thread has been stopped, make sure it has been joined
+    if (m_state == Stopped) {
+        if (m_thread.joinable()) {
+            lock.unlock();
+            m_thread.join();
+            lock.lock();
+        }
+    }
+
+    m_logger_ptr->info("Event loop start()");
+    m_state = Started;
+    auto start_pf = util::make_promise_future<void>();
```

Comment:
If we're going to track the state of the thread, do we need separate futures for starting/stopping events? Can we just add a condvar that you wait on for m_state to change?